### ISSUE TITLE
Buying items from traders not removing iConomy currency?

### ISSUE THREAD

**Issue by nmerchant** [CitizensDev/Citizens#743]
Someone on my server noticed this yesterday - I had an NPC set up to buy / sell diamond.  He bought several stacks and noticed his money did not deplete, so he would sell them all back to make an enormous profit, buy some more for free, repeat.  Using latest Citizens and iConomy (although it hasn't been updated in a while it seems to be fully functional, except for this).  Is this a problem with Citizens or is another economy mod compatible?

**Comment by Nihilhalla** [comment]
im having this issue as well, and can confirm that i have the latest citizens and iconomy

Edit: i do also have the RB1.1R3 jar, for craftbukkit, and if i had a suspicion as to what is causing the issue, i would have to say its an
Iconomy error, for example, using the function for Iconomy to give money, rather than pay money,

/money give <name> adds funds to that account, /money pay <name> deducts that money from the issuer's account to add funds to the target account

**Comment by zackb1079** [comment]
I am also having this issue. Running on Bukkit build 1.1R3. Players may sell items for money but when buying items no money is deducted.

**Comment by Nihilhalla** [comment]
bump

**Comment by JaredTS486** [comment]
I am running craftbukkit revision 4 and the latest 1.1.5 citizens and iconomy. iConomy is running in a mysql database and citizens traders are indeed editing the money from their when buying / repairing.

**Comment by Nihilhalla** [comment]
so the issue is a bukkit method that fell by the wayside?

i will test to confirm

**Comment by Nihilhalla** [comment]
looks like R4 is actually not a RB, and it would break some of the other plugins i have on my server, so i will wait to confirm

**Comment by JaredTS486** [comment]
R4 shouldnt break any mods, but i did try with the R3 and iConomy is still working correctly.

**Comment by Nihilhalla** [comment]
what other plugins are you using? spout?

i have spout, citizens, multiverse, iConomy, PEX, and towny. those are the only plugins that would modify anything to do with iconomy

**Comment by Nihilhalla** [comment]
My other plugins that utilize Iconomy have been fully tested, they all produce the desired results, citizens is the only malfunctioning plugin, im gonna try reinstalling it, just in case it somehow rewrote itself. i am at a full loss at this point,

**Comment by JaredTS486** [comment]
The only plugins I don't have are townhouse and multiverse. It wouldn't
hurt to try running the server without those to see if that makes a
difference.
On Feb 6, 2012 2:11 AM, "Rainemakerx" <
reply@reply.github.com>
wrote:

**Comment by nmerchant** [comment]
Update:

I don't know what I did but somehow this is working.  I had removed all my plugins temporarily (renamed folder to __plugins) to test a single plugin by itself.  This somehow reset my config file, as when I moved them back, creating NPCs / using teleporters costed money again.  And now, NPCs now properly take people's money.

**Comment by Nihilhalla** [comment]
I have tested this multiple ways with the R3 jar, without citizens then with it again, wiping the mysql tables for iConomy (shouldnt have worked anyway seeing as all my other iConomy dependant plugins are fine), deleted all my configs and let citizens recreate them. I fail to see how none of this could have helped if it was not a code issue, the only thing I can think is that vault is part of the issue, but vault really shouldnt have any effect on iConomy or citizens, as the only plugin I have installed that uses it is Towny.

**Comment by JaredTS486** [comment]
All I can say is I have iconomy set up using MySQL. It may not be writing
or reading from the mini database file correctly.
On Feb 8, 2012 12:26 AM, "Rainemakerx" <
reply@reply.github.com>
wrote:

**Comment by JaredTS486** [comment]
Are you using essentials?

**Comment by Nihilhalla** [comment]
ive never used essentials, what minidb are you referring to? i have MySQL for iConomy, and citizens is flatfile only, and will be until citizens 2

**Comment by JaredTS486** [comment]
Can you post a server log using pasties? Just from start to finish when
loading the server.
On Feb 9, 2012 2:34 AM, "Rainemakerx" <
reply@reply.github.com>
wrote:

**Comment by Nihilhalla** [comment]
http://pastebin.com/raw.php?i=n92tYaYc

Edit: just realized register was still in there, i wasnt certain if towny still needed it or not, even still that shouldn't be causing the issue

**Comment by JaredTS486** [comment]
I just started getting this issue as well. I was looking to find out what plugins were causing issues with ShowcaseStandalone. Now they are not removing the currency from iConomy. However the ONLY plugin thats new in my server that we also have in common is Catacombs. Im going to try and remove it later today to see if that works. ALSO i would try to disable register and see if that works.

**Comment by Nihilhalla** [comment]
I added catatcombs after the fact, im gonna try disabling register.

**Comment by Nihilhalla** [comment]
disabled register, all it did was kill town banks and plot claim costs, reenabled it.

**Comment by Nihilhalla** [comment]
made a bare server with nothing but spout, pex iconomy and citizens, same issue, the only thing i can think (since spout has not been an issue before) is that citizens is encountering an illegal character on the sql tables, but wouldnt that make iconomy and other dependent plugins crash also?

**Comment by JaredTS486** [comment]
As far as I can tell iconomy is hooking with citizens right. Its something
else. So I'm having an issue with my blacksmith. He is repairing things for
free. Is your blacksmith saying its repair the item for blank money but not
updating the currency?
On Feb 10, 2012 8:59 PM, "Rainemakerx" <
reply@reply.github.com>
wrote:

**Comment by Nihilhalla** [comment]
no, its saying repaired your (item). no currency mentioned at all

**Comment by Nihilhalla** [comment]
any thoughts Jared? its as though economy has been forcibly disabled across the entire citizens plugin, thats the only issue im getting, yours appears to be localized to the blacksmith jar, on that i have no clue, as obviously i cant even fix my own problem

**Comment by Nihilhalla** [comment]
nmerchant, are you using mysql for your iconomy db?

**Comment by HiddenLurker** [comment]
I can also confirm this issue. That's why I didn't let my trader sell anything.

All my friends were wondering why we could sell all sorts of things to the trader but not buy anything from him.

Any solutions so far?

Sucks not to be able to buy anything with your hard-earned money.

**Comment by JaredTS486** [comment]
Now that I look you you can sell things but not buy from the traders on my
site as well. So we know its just when spending the money. And I have
essentials and showcase standalone working just fine on my server money
updated fine. Things that mess with iconomy are permissions and vault for
hooks. But on another note. We should try to run citizens and iconomy as
the only plugins on the server. I can't ATM, I have lots of members on. I'm
going to set up a test server on my laptop to try and debug with. Also are
you guys using mcmyadmin?
On Feb 12, 2012 6:46 AM, "HiddenLurker" <
reply@reply.github.com>
wrote:

**Comment by Nihilhalla** [comment]
i can bring a test server online if you like, but i did try once with just citizens, spout, and iconomy with the same results as my current server. let me try with just iconomy and citizens

Edit: just got the bit about mcmyadmin, and negative, but we do have an ssh into the box

**Comment by JaredTS486** [comment]
New version of citizens and latest version of iconomy works on my test
server. Going to update citizens. New version is 16

**Comment by Nihilhalla** [comment]
so the apparent issue is with spout then?

Edit: test server without spout reports success, but so does test server with spout. going to try dropping the Iconomy DB and remaking it on the sql tables, be back soon with results

Edit2: well, i attached it to the old ico db, and it still worked, im gonna start adding iconomy hooked plugins until it breaks to find the culprit, also i found a neat bug with blacksmiths, they use their own money to repair, so you have to give them the money beforehand in order to repair,

Edit3: in regard to the additional error, it actually takes the sum from both parties, so two people are getting charged for the repair, the npc and the player....

Edit4: though it pains me to say it, PEX is what is breaking the iconomy-citizens relationship, i have to admit i did not see that coming at all, but now that we know what it is, we can figure out how to fix it

**Comment by Nihilhalla** [comment]
@fullwall  sorry about my outburst on issue 794 and the doubt i have placed on you, issue 743 has been tested by adding one plugin at a time until the issue was able to be recreated, the plugin that allowed for the recreation of the bug was Permissions EX, still odd though, considering that it updated and that it was working until 1.1.5 citizens came out

**Comment by Nihilhalla** [comment]
Initial findings suggested a missing permission in the default group, that was preventing the npc from taking my money, or indeed modifying my balance at all. however after giving the test server's default group the '*' permission i noticed no change in the behavior, this would indicate an error between citizens and PEX on the hook level, or perhaps there is a permission that cannot be given?

**Comment by nmerchant** [comment]
@Rainemakerx no I am not.  If you'd like I could zip and upload my plugins folder somewhere and you could take a look.  As far as I know buying is still working since I did that weird config reset that seemed to have fixed it.

**Comment by Nihilhalla** [comment]
if you have PEX in there, that seems to be the root of the cause, if not we have to look at other plugins again, or combinations of plugins.

**Comment by nmerchant** [comment]
Yup, I have PEX, iConomy, Citizens, MagicSpells, Factions.

**Comment by Nihilhalla** [comment]
Ok, Try making a test server without PEX and tell me what happens. i did this myself but i want to confirm that it is JUST PEX causing the issue

**Comment by Nihilhalla** [comment]
backup your current server if you opt to make your primary the test server

**Comment by JaredTS486** [comment]
I know I sound like a broken record but I still think this issue can't be
isolated with pex. Are there any other permissions mods that work with
citizens? Once apon a time PEx wasn't having this issue with citizens. I
read in another post removing modifyworld.jar fixes this. That might be
worth looking into.
On Feb 13, 2012 10:13 AM, "Rainemakerx" <
reply@reply.github.com>
wrote:
